rl4co/envs/scheduling/jssp.py

```diff
@@ -0,0 +1,665 @@
+from typing import Optional, Tuple
+
+import torch
+
+from tensordict.tensordict import TensorDict
+from torchrl.data import CompositeSpec, DiscreteTensorSpec, UnboundedContinuousTensorSpec
+
+from rl4co.envs.common.base import RL4COEnvBase
+
+
+class Configs:
+    low = 1
+    high = 99
+    et_normalize_coef = 1000
+    rewardscale = 0.0
+    init_quality_flag = False
+
+
+configs = Configs()
+
+
+class JSSPEnv(RL4COEnvBase):
+    """Job Shop Scheduling Problem (JSSP) environment.
+    As per the definition given in https://arxiv.org/pdf/2010.12367.pdf.
+    The goal is to schedule a set of jobs on a set of machines such that the makespan is minimized.
+    In this variation, the number of operations per job is equal to the number of machines.
+
+    Args:
+
+    Note:
+        -
+    """
+
+    name = "jssp"
+
+    def __init__(self, num_jobs, num_machines, **kwargs):
+        super().__init__(**kwargs)
+        self.batch_size = torch.Size([1])
+        adjacency_spec = DiscreteTensorSpec(
+            n=2,
+            shape=torch.Size((1, num_jobs * num_machines, num_jobs * num_machines)),
+            device=self.device,
+            dtype=torch.int64,
+        )
+        features_spec = UnboundedContinuousTensorSpec(
+            shape=torch.Size((1, num_jobs * num_machines, 2)),
+            device=self.device,
+        )
+        feasible_actions_spec = DiscreteTensorSpec(
+            n=num_jobs * num_machines,
+            shape=torch.Size(
+                (
+                    1,
+                    num_jobs,
+                )
+            ),
+            device=self.device,
+            dtype=torch.int64,
+        )
+        action_mask_spec = DiscreteTensorSpec(
+            n=2,
+            shape=torch.Size(
+                (
+                    1,
+                    num_jobs,
+                )
+            ),
+            device=self.device,
+            dtype=torch.bool,
+        )
+
+        self.observation_spec = CompositeSpec(
+            adjacency=adjacency_spec,
+            features=features_spec,
+            feasible_actions=feasible_actions_spec,
+            action_mask=action_mask_spec,
+            shape=self.batch_size,
+        )
+
+        self.action_spec = DiscreteTensorSpec(
+            n=num_jobs,
+            shape=self.batch_size,
+            device=self.device,
+            dtype=torch.int64,
+        )
+
+        self.num_jobs = num_jobs
+        self.num_machines = num_machines
+        self.num_tasks = self.num_jobs * self.num_machines
+        # the task id for first column
+        self.first_col = torch.arange(
```

Comment:
Here it seems that there are some tensors declared inside the  `__init__` signature; this makes the environment _stateful_ (as explained [here](https://pytorch.org/rl/tutorials/pendulum.html#:~:text=To%20give%20a,API%20in%20TorchRL.)) .

I don't hold a grudge against stateful envs, but I wonder whether it would be better (i.e. more efficient) to have them as in the [pendulum example](https://pytorch.org/rl/tutorials/pendulum.html), which is, simply passed upon generation?